File: lib/pageflow/react/widget_type.rb
module Pageflow
  module React
    class WidgetType < Pageflow::WidgetType
      attr_reader :name, :role, :options

      def initialize(name, role, options = {})
        @name = name
        @role = role
        @options = options
      end

      def roles
        [role]
      end

      def insert_point
        @options[:insert_point] || super
      end

      def render(template, _)
        template.render(File.join('pageflow', 'react', 'widget'), name: name)
      end
    end
  end
end

File: package/src/editor/api/WidgetType.js
import _ from 'underscore';
import {Object} from 'pageflow/ui';

export const WidgetType = Object.extend({
  initialize: function(serverSideConfig, clientSideConfig) {
    this.name = serverSideConfig.name;
    this.translationKey = serverSideConfig.translationKey;
    this.insertPoint = serverSideConfig.insertPoint;
    this.configurationEditorView = clientSideConfig.configurationEditorView;
    this.configurationEditorTabViewGroups = clientSideConfig.configurationEditorTabViewGroups || {};
    this.isOptional = clientSideConfig.isOptional;
  },

  hasConfiguration: function() {
    return !!this.configurationEditorView;
  },

  createConfigurationEditorView: function(options) {
    var constructor = this.configurationEditorView;
    return new constructor(_.extend({
      attributeTranslationKeyPrefixes: [
        'pageflow.editor.widgets.attributes.' + this.name,
        'pageflow.editor.widgets.common_attributes'
      ]
    }, options));
  },

  defineStubConfigurationEditorTabViewGroups(groups) {
    _.each(this.configurationEditorTabViewGroups, (fn, name) =>
      groups.define(name, () => {})
    );
  },

  defineConfigurationEditorTabViewGroups(groups) {
    _.each(this.configurationEditorTabViewGroups, (fn, name) =>
      groups.define(name, fn)
    );
  }
});
